### ISSUE TITLE
Assertion failed: (last_ref), function ~ColumnFamilySet, file /repos/rocksdb/db/column_family.cc

### ISSUE THREAD

**Issue by toktarev** [facebook/rocksdb#4063]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://www.facebook.com/groups/rocksdb.dev

RocksDB version: 5.13.2

### Expected behavior

Infinite program work

### Actual behavior
Assertion failed: (last_ref), function ~ColumnFamilySet, file /repos/rocksdb/db/column_family.cc
line 1166
### Steps to reproduce the behavior

```
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

#include <cstdio>
#include <string>
#include <iostream>
#include <rocksdb/slice_transform.h>
#include <rocksdb/table.h>
#include "rocksdb/db.h"
#include "rocksdb/stacktrace.h"
#include <rocksdb/filter_policy.h>
#include <thread>
#include <cinttypes>
#include "starrockslistener.h"

using namespace std;
using namespace rocksdb;

using namespace std::chrono;

std::string kDBPath = "/repos/rocksdata";

template<typename T>
T swap_endian(T u) {
    static_assert(CHAR_BIT == 8, "CHAR_BIT != 8");

    union {
        T u;
        unsigned char u8[sizeof(T)];
    } source, dest;

    source.u = u;

    for (size_t k = 0; k < sizeof(T); k++)
        dest.u8[k] = source.u8[sizeof(T) - k - 1];

    return dest.u;
}

rocksdb::TableFactory *makeDictionaryTableFactory2() {
    auto blockOpts = rocksdb::BlockBasedTableOptions{};
    blockOpts.index_type = BlockBasedTableOptions::IndexType::kTwoLevelIndexSearch;
    blockOpts.partition_filters = true;
    blockOpts.metadata_block_size = 4096;
    blockOpts.block_size = 64ul << 10u;
    blockOpts.block_cache = NewLRUCache(static_cast<size_t>(1024 * 1024 * 1024));
    blockOpts.cache_index_and_filter_blocks = false;
    blockOpts.cache_index_and_filter_blocks_with_high_priority = true;

    auto *pPolicy = rocksdb::NewBloomFilterPolicy(10, false);
    auto filter_ptr = std::shared_ptr<const rocksdb::FilterPolicy>(pPolicy);
    blockOpts.filter_policy = filter_ptr;

    return rocksdb::NewBlockBasedTableFactory(blockOpts);
}


void process(ColumnFamilyHandle *cf, DB *db, u_int64_t prefix) {
    uint64_t start = (uint64_t) std::chrono::duration_cast<std::chrono::seconds>(
            system_clock::now().time_since_epoch()).count();

    Status s;
    u_int64_t *buffer = new u_int64_t[4];
    char *pointer = reinterpret_cast<char *>(buffer);
    WriteBatch writeBatch{};
    u_int64_t max = 0;

    Slice key(pointer, 32);
    Slice value(reinterpret_cast<char *>(&max), 8);

    u_int64_t aNumber = 1000000;

    for (u_int64_t aIndex = aNumber * prefix; aIndex < (aNumber * prefix) + aNumber; aIndex++) {
        *(buffer) = swap_endian(aIndex);
        *(buffer + 1) = aIndex + 1;
        *(buffer + 2) = aIndex + 2;
        *(buffer + 3) = aIndex + 3;

        writeBatch.Put(cf, key, value);

        if (aIndex % 1000 == 0) {
            Status ss = db->Write(WriteOptions(), &writeBatch);
            assert(ss.ok());
            writeBatch.Clear();
        }
    }

    db->Flush(FlushOptions(), cf);

    uint64_t end = (uint64_t) std::chrono::duration_cast<std::chrono::seconds>(
            system_clock::now().time_since_epoch()).count();
    u_int64_t delta = end - start;

    printf("%" PRIu64 "\n", delta);
}

void initiate() {
    system("rm -rf /repos/rocksdata/*");

    DB *db;

    rocksdb::Options options{};
    options.create_if_missing = true;
    options.compression = CompressionType::kNoCompression;

    options.avoid_flush_during_shutdown = true;

    options.num_levels = 2;
    options.create_if_missing = true;
    options.create_missing_column_families = true;
    options.use_adaptive_mutex = true;

    options.IncreaseParallelism(std::thread::hardware_concurrency());
    options.OptimizeLevelStyleCompaction();

    for (int i = 0; i < 2; ++i) {
        options.compression_per_level[i] = kNoCompression;
    }

    options.disable_auto_compactions = true;

    auto *tf = makeDictionaryTableFactory2();
    options.table_factory =
            std::shared_ptr<rocksdb::TableFactory>(tf);

    options.write_buffer_size = 128 * 1024 * 1024;
    options.allow_concurrent_memtable_write = true;

    options.error_if_exists = false;

    options.max_background_compactions = 1;
    options.max_background_flushes = 1;

    std::vector<ColumnFamilyHandle *> handles;
    Status s = DB::Open(options, kDBPath, &db);

    if (!s.ok()) {
        std::cout << s.ToString();
    }

    assert(s.ok());

    ColumnFamilyOptions cf_options{};

    cf_options.table_factory.reset(makeDictionaryTableFactory2());
    cf_options.prefix_extractor.reset(rocksdb::NewNoopTransform());

    ColumnFamilyHandle *cf;
    std::string name("Name");
    Status status = db->CreateColumnFamily(cf_options, name, &cf);
    assert(s.ok());

    thread t1(process, cf, db, 1);
    thread t2(process, cf, db, 2);
    t1.join();
    t2.join();

    db->DestroyColumnFamilyHandle(cf);

    delete db;
}

int main() {
    initiate();

    while (true) {
        DB *db;

        rocksdb::Options options{};
        options.create_if_missing = true;
        options.compression = CompressionType::kNoCompression;

        options.avoid_flush_during_shutdown = true;

        options.num_levels = 2;
        options.create_if_missing = true;
        options.create_missing_column_families = true;
        options.use_adaptive_mutex = true;

        options.IncreaseParallelism(std::thread::hardware_concurrency());
        options.OptimizeLevelStyleCompaction();

        for (int i = 0; i < 2; ++i) {
            options.compression_per_level[i] = kNoCompression;
        }

        options.disable_auto_compactions = true;

        auto *tf = makeDictionaryTableFactory2();
        options.table_factory =
                std::shared_ptr<rocksdb::TableFactory>(tf);

        options.write_buffer_size = 128 * 1024 * 1024;
        options.allow_concurrent_memtable_write = true;

        options.error_if_exists = false;

        options.max_background_compactions = 1;
        options.max_background_flushes = 1;

        std::vector<ColumnFamilyDescriptor> column_families;
        // have to open default column family
        column_families.push_back(ColumnFamilyDescriptor(
                "default", ColumnFamilyOptions()));
        // open the new one, too
        column_families.push_back(ColumnFamilyDescriptor(
                "Name", ColumnFamilyOptions()));
        std::vector<ColumnFamilyHandle *> handles;
        Status s = DB::Open(options, kDBPath, column_families, &handles, &db);

        if (!s.ok()) {
            std::cout << s.ToString();
        }

        assert(s.ok());

        ColumnFamilyOptions cf_options{};

        cf_options.table_factory.reset(makeDictionaryTableFactory2());
        cf_options.prefix_extractor.reset(rocksdb::NewNoopTransform());

        ColumnFamilyHandle *cf = handles[1];
        assert(s.ok());

        thread t1(process, cf, db, 1);
        thread t2(process, cf, db, 2);
        t1.join();
        t2.join();

        db->DestroyColumnFamilyHandle(cf);

        delete db;
    }
}


```

**Comment by guoxiangCN** [comment]
Hello, I also encountered this problem. How did you solve it ？

**Comment by taiyang-li** [comment]
Me too. Any progress ?

**Comment by linas** [comment]
I'm seeing this problem in the current Debian testing `librocksdb-dev/testing,now 7.2.2-5 amd64` https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1016394 I'm going to try to reproduce on git master and see what happens ...

**Comment by linas** [comment]
I just opened #10474 to track today's version of this bug.

**Comment by cbi42** [comment]
In the given code example, I think `DestroyColumnFamilyHandle` needs to be called on all handle in `handles`.

**Comment by cocalele** [comment]
> In the given code example, I think `DestroyColumnFamilyHandle` needs to be called on all handle in `handles`.

In my case https://github.com/cocalele/ViveNAS/issues/8, destroy all column family, include default cf, solves this problem.

**Comment by ottok** [comment]
I am the maintainer of RocksDB in Debian and I am trying to follow up if https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1016394 was fixed. Both this #4063 and #10474 are marked as closed, but neither has a commit linked or any explanation on _why_ the issues were closed, so wondering what the status is.

**Comment by linas** [comment]
Hi @ottok I think this is closed because its "user error".  In my case, I saw this error because I had failed to delete rockdb iterators after I was done with them. This resulted in a memory leak, and eventually hitting this error (after the leak got big enough). Older versions of rocksdb just leaked silently; newer versions actually issued a warning/error (where "older" and "newer" were "whatever debian stable had in 2022 when I commented" and "whatever was in the rocksdb git tree in 2022")

Once I figured out it was my own fault, for leaking iterators, I closed the issue that I had created. I figured this out in about a week.

The person who opened this issue @toktarev opened and closed it on the same day. I think that means  he also realized it was "user error", but did not explain what the problem was.  And so everyone who makes this mistake ends up here, scratching their head.

The person who opened https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1016394  was me. When I finally realized it was "user error", I forgot to close the Debian bug report. Ooops. I'll try to do that now.

**Comment by jonassmedegaard** [comment]
According to #13106 it seems some memory leaks are caused not by user error but bugs in internal cleanup routines, partly fixed in v9.8.4 but partly fixed only in v9.10.0.
This seems to be the case for Debian [bug #1088523](https://bugs.debian.org/1088523), which oddly seems [unreproducible with a pristine build of rocksdb v0.8.4](https://github.com/oxigraph/oxigraph/issues/1136#issuecomment-2567543839).
